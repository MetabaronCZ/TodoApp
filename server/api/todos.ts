import { Application } from 'express';
import { InferType, boolean, object, string } from 'yup';

import { Todo } from 'models/Todo';

import { mockedDb } from '../db';
import {
  mockApiRequest,
  mockCreatedId,
  mockParseParamArray,
  mockParseParam,
  mockParseInteger,
} from '../utils';

const createTodoDataSchema = object({
  title: string().required(),
  description: string().required(),
  folder: string().nullable().required(),
  isDone: boolean().required(),
}).required();

const editTodoDataSchema = object({
  title: string(),
  description: string(),
  folder: string().nullable(),
  isDone: boolean(),
}).required();

export const setApiTodoEndpoints = (app: Application): void => {
  // get todo list
  app.get('/todo', async (req, res) => {
    await mockApiRequest();

    // parse params
    const perPage = mockParseInteger(req.query.perPage) ?? 20;
    const page = mockParseInteger(req.query.page) ?? 0;
    const folder = mockParseParam(req.query.folder);
    const query = mockParseParam(req.query.query);
    const sort = mockParseParam(req.query.sort);

    // filter items
    const items = mockedDb.todos.filter((item) => {
      // filter by folder
      if (folder && folder !== item.folder) {
        return false;
      }
      // filter by query string
      if (
        query &&
        !item.title.toLowerCase().includes(query.toLowerCase()) &&
        !item.description.toLowerCase().includes(query.toLowerCase())
      ) {
        return false;
      }
      return true;
    });

    // sort items
    items.sort((a, b) => {
      switch (sort) {
        case 'TITLE_ASC':
          return a.title.localeCompare(b.title);
        case 'TITLE_DESC':
          return b.title.localeCompare(a.title);
        case 'CREATED_ASC':
          return a.created - b.created;
        case 'CREATED_DESC':
          return b.created - a.created;
        case 'DONE_1':
          return a.isDone === b.isDone ? 0 : a.isDone ? -1 : +1;
        case 'DONE_0':
          return a.isDone === b.isDone ? 0 : a.isDone ? +1 : -1;
        default:
          // use CREATED_DESC as default
          return b.created - a.created;
      }
    });

    const start = page * perPage;
    const end = start + perPage;

    res.json({
      data: {
        items: items.slice(start, end),
        count: items.length,
      },
    });
  });

  // get todo detail
  app.get('/todo/:id', async (req, res) => {
    const { id } = req.params;
    await mockApiRequest();

    const todo = mockedDb.todos.find((item) => id === item.id);
    res.json({ data: todo || null });
  });

  // create todo
  app.post('/todo', async (req, res) => {
    const data = req.body;
    await mockApiRequest();

    try {
      await createTodoDataSchema.validate(data);
      const todoData = data as InferType<typeof createTodoDataSchema>;

      const todo: Todo = {
        ...todoData,
        id: mockCreatedId('TODO'),
        created: Date.now(),
      };
      mockedDb.todos = [...mockedDb.todos, todo];
      res.json({ data: todo });
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // edit todo
  app.patch('/todo/:id', async (req, res) => {
    const { id } = req.params;
    const data = req.body;
    await mockApiRequest();

    try {
      await editTodoDataSchema.validate(data);
      const todoData = data as InferType<typeof editTodoDataSchema>;

      mockedDb.todos = mockedDb.todos.map((item) => {
        return id === item.id ? { ...item, ...todoData } : item;
      });

      res.json(null);
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // delete todos
  app.delete('/todo', async (req, res) => {
    const { ids } = req.query;
    await mockApiRequest();

    const idsParsed = mockParseParamArray(ids);

    const validIds = mockedDb.todos
      .filter((item) => idsParsed.includes(item.id))
      .map((item) => item.id);

    if (0 === validIds.length) {
      res.status(500).json({ error: 'Could not delete todos: Invalid ID!' });
    } else {
      mockedDb.todos = mockedDb.todos.filter((item) => {
        return !validIds.includes(item.id);
      });

      res.json(null);
    }
  });
};
