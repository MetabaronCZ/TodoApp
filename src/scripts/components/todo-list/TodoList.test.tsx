import React from 'react';
import { t } from 'i18next';
import { Provider } from 'react-redux';

import { act, render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import { TodoList } from 'components/todo-list/TodoList';
import * as TodoListItem from 'components/todo-list/TodoListItem';
import * as TodoListToolbar from 'components/todo-list/TodoListToolbar';

import { client } from 'client';
import { Todo } from 'models/Todo';
import { todoListSlice } from 'store/todos';

import { mockStore } from 'test/store';
import { withMockedProviders } from 'test/component';

const { setTodos } = todoListSlice.actions;

const testData: Todo[] = [
  {
    id: 'A',
    title: 'Todo A',
    description: '',
    created: 0,
    folder: null,
    isDone: false,
  },
  {
    id: 'B',
    title: 'Todo B',
    description: '',
    created: 1,
    folder: '0',
    isDone: true,
  },
  {
    id: 'C',
    title: 'Todo B',
    description: '',
    created: 2,
    folder: null,
    isDone: false,
  },
];

const getTodoList = (todos: Todo[]): JSX.Element => {
  const store = mockStore();
  store.dispatch(setTodos(todos));

  return withMockedProviders(
    <Provider store={store}>
      <TodoList />
    </Provider>,
  );
};

// get component props for last "n" mocked calls
const getItemsProps = <T extends object>(
  component: jest.SpiedFunction<React.FC<T>>,
  lastCalls = 3,
): Array<T> => {
  const { calls } = component.mock;
  return calls.map((call) => call[0]).slice(-lastCalls);
};

describe('components/todo-list/TodoList', () => {
  it('should render correctly', () => {
    const listItem = jest.spyOn(TodoListItem, 'TodoListItem');
    expect(listItem).toBeCalledTimes(0);

    const tree = render(getTodoList(testData));
    expect(tree.container).toMatchSnapshot();

    const { calls } = listItem.mock;
    const ids = calls.map((call) => call[0].item.id);
    expect(ids.length % 3).toEqual(0); // check item count (with re-renders)
    expect(ids.at(-3)).toEqual('A');
    expect(ids.at(-2)).toEqual('B');
    expect(ids.at(-1)).toEqual('C');
  });

  it('should render empty list', () => {
    const listItem = jest.spyOn(TodoListItem, 'TodoListItem');
    const toolbar = jest.spyOn(TodoListToolbar, 'TodoListToolbar');
    expect(listItem).toBeCalledTimes(0);
    expect(toolbar).toBeCalledTimes(0);

    const tree = render(getTodoList([]));
    expect(tree.container).toMatchSnapshot();

    expect(listItem).toBeCalledTimes(0);
    expect(tree.getByText(t('todoList.empty'))).toBeInTheDocument();

    expect(toolbar).toHaveBeenCalled();
    expect(toolbar).lastCalledWith(
      expect.objectContaining({ disabled: true }),
      expect.anything(),
    );
  });

  it('should be selectable', () => {
    const listItem = jest.spyOn(TodoListItem, 'TodoListItem');
    const toolbar = jest.spyOn(TodoListToolbar, 'TodoListToolbar');

    render(getTodoList(testData));

    let items = getItemsProps(listItem);
    expect(items[0].selected).toEqual(false);
    expect(items[1].selected).toEqual(false);
    expect(items[2].selected).toEqual(false);

    expect(toolbar).toHaveBeenCalled();

    if (toolbar.mock.lastCall) {
      expect(toolbar.mock.lastCall[0].selected).toEqual(false);
    }
    toolbar.mockClear();

    const testSelectItem = (
      index: 0 | 1 | 2,
      shouldSelect: boolean,
      expectation: [boolean, boolean, boolean],
      selectedAll: boolean,
    ): void => {
      act(() => items[index].onSelect(shouldSelect));

      items = getItemsProps(listItem);
      expect(items[0].selected).toEqual(expectation[0]);
      expect(items[1].selected).toEqual(expectation[1]);
      expect(items[2].selected).toEqual(expectation[2]);

      expect(toolbar).toHaveBeenCalled();

      if (toolbar.mock.lastCall) {
        // select-all checkbox in toolbar active
        expect(toolbar.mock.lastCall[0].selected).toEqual(selectedAll);
      }
      toolbar.mockClear();
    };

    testSelectItem(1, true, [false, true, false], true); // select 2nd item
    testSelectItem(0, true, [true, true, false], true); // select 1st item
    testSelectItem(1, false, [true, false, false], true); // deselect 2nd item
  });

  it('should be able to un/select-all', () => {
    const listItem = jest.spyOn(TodoListItem, 'TodoListItem');
    const toolbar = jest.spyOn(TodoListToolbar, 'TodoListToolbar');

    render(getTodoList(testData));

    let items = getItemsProps(listItem);
    expect(items[0].selected).toEqual(false);
    expect(items[1].selected).toEqual(false);
    expect(items[2].selected).toEqual(false);

    expect(toolbar).toHaveBeenCalled();

    // trigger select all
    act(() => {
      if (toolbar.mock.lastCall) {
        toolbar.mock.lastCall[0].onSelect(true);
      }
    });

    items = getItemsProps(listItem);
    expect(items[0].selected).toEqual(true);
    expect(items[1].selected).toEqual(true);
    expect(items[2].selected).toEqual(true);

    // trigger deselect all
    act(() => {
      if (toolbar.mock.lastCall) {
        toolbar.mock.lastCall[0].onSelect(false);
      }
    });

    items = getItemsProps(listItem);
    expect(items[0].selected).toEqual(false);
    expect(items[1].selected).toEqual(false);
    expect(items[2].selected).toEqual(false);
  });

  it('should be sortable', () => {
    const toolbar = jest.spyOn(TodoListToolbar, 'TodoListToolbar');

    const mockedSortTodos = jest.spyOn(client.todo, 'get');
    expect(mockedSortTodos).toBeCalledTimes(0);

    render(getTodoList(testData));
    expect(toolbar).toHaveBeenCalled();

    if (toolbar.mock.lastCall) {
      expect(toolbar.mock.lastCall[0].sort).toEqual('CREATED_DESC');
    }

    // trigger todo sort
    act(() => {
      if (toolbar.mock.lastCall) {
        toolbar.mock.lastCall[0].onSort('TITLE_DESC');
      }
    });

    expect(mockedSortTodos).toBeCalledTimes(1);

    if (mockedSortTodos.mock.lastCall) {
      expect(mockedSortTodos.mock.lastCall[0]?.sort).toEqual('TITLE_DESC');
    }
  });

  it('should be able to delete selected', () => {
    const listItem = jest.spyOn(TodoListItem, 'TodoListItem');
    const toolbar = jest.spyOn(TodoListToolbar, 'TodoListToolbar');

    const mockedConfirm = jest.spyOn(window, 'confirm');
    mockedConfirm.mockImplementation(jest.fn(() => true));
    expect(mockedConfirm).toBeCalledTimes(0);

    const mockedDeleteTodos = jest.spyOn(client.todo, 'delete');
    expect(mockedDeleteTodos).toBeCalledTimes(0);

    render(getTodoList(testData));

    let items = getItemsProps(listItem);

    // select first two items
    act(() => {
      items[0].onSelect(true);
      items[1].onSelect(true);
    });

    items = getItemsProps(listItem);
    expect(items[0].selected).toEqual(true);
    expect(items[1].selected).toEqual(true);
    expect(items[2].selected).toEqual(false);

    expect(toolbar).toHaveBeenCalled();

    // trigger toolbar onDelete
    act(() => {
      if (toolbar.mock.lastCall) {
        toolbar.mock.lastCall[0].onDelete();
      }
    });

    expect(mockedConfirm).toBeCalledTimes(1);
    expect(mockedDeleteTodos).toBeCalledTimes(1);
    expect(mockedDeleteTodos).toBeCalledWith(['A', 'B']);
  });
});
