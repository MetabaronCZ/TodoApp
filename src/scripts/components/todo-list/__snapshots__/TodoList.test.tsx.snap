// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/todo-list/TodoList should render correctly 1`] = `
.c9 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c12 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c13 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c13:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c13:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c13:disabled {
  color: #c0c0c0;
}

.c20 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
}

.c20:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c20:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c20:disabled {
  color: #c0c0c0;
}

.c8 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c8:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c8:disabled {
  color: #c0c0c0;
}

.c10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 18px;
  line-height: 30px;
  font-weight: normal;
  margin: 0;
}

.c14 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c15 {
  border-bottom: 1px solid #c0c0c0;
}

.c15:last-child {
  border-bottom: none;
}

.c6 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: 30px;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  width: 20px;
  height: 20px;
  border: 2px solid #222;
  color: #222;
  background: #fff;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c4:hover {
  background-color: #fcfcfc;
}

.c4:focus-within {
  outline: 1px dotted #222;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
  opacity: 0;
  outline: none;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
}

.c17 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 5px 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c19 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  width: 20px;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c18 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  white-space: nowrap;
}

.c2 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: calc(30px + 1px);
  border-bottom: 1px solid #c0c0c0;
  background: #f0f0f0;
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c11 {
  position: relative;
  display: inline-block;
}

.c1 {
  margin-bottom: 1px;
}

<div>
  <div
    class=""
  >
    <h1
      class="c0 c1"
    >
      Úkoly
    </h1>
    <div
      class="c2"
    >
      <div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="checkbox-4"
          >
            <input
              class="c5"
              id="checkbox-4"
              type="checkbox"
            />
          </label>
          <label
            class="c6"
            for="checkbox-4"
          >
            Vybrat vše
          </label>
        </div>
      </div>
      <div
        class="c7"
      />
      <div>
        <a
          class="c8"
          href="/todo/create"
        >
          <i
            class="c9"
          >
            +
          </i>
          <span
            class="c10"
          >
            Vytvořit
          </span>
        </a>
      </div>
      <div>
        <div
          class="c11"
        >
          <button
            class="c12 c13"
            type="button"
          >
            <span
              class="c10"
            >
              Nejnovější
            </span>
            <i
              class="c9"
            >
              ˅
            </i>
          </button>
        </div>
      </div>
    </div>
    <ul
      class="c14"
    >
      <li
        class="c15"
      >
        <div
          class="c16"
        >
          <div
            class=""
          >
            <div
              class="c3"
            >
              <label
                class="c4"
                for="checkbox-5"
              >
                <input
                  class="c5"
                  id="checkbox-5"
                  type="checkbox"
                />
              </label>
            </div>
          </div>
          <a
            class="c17"
            href="/todo/A"
          >
            Todo A
          </a>
          <div
            class="c18"
          >
            1. 1. 1970
          </div>
          <div
            class="c19"
            title=""
          >
            <i
              class="c9"
            >
              -
            </i>
          </div>
          <div
            class=""
          >
            <a
              class="c8"
              href="/todo/A"
            >
              <i
                class="c9"
              >
                ✎
              </i>
              <span
                class="c10"
              >
                Upravit
              </span>
            </a>
          </div>
          <div
            class=""
          >
            <button
              class="c12 c20"
              type="button"
            >
              <i
                class="c9"
              >
                ✖
              </i>
              <span
                class="c10"
              >
                Smazat
              </span>
            </button>
          </div>
        </div>
      </li>
      <li
        class="c15"
      >
        <div
          class="c16"
        >
          <div
            class=""
          >
            <div
              class="c3"
            >
              <label
                class="c4"
                for="checkbox-6"
              >
                <input
                  class="c5"
                  id="checkbox-6"
                  type="checkbox"
                />
              </label>
            </div>
          </div>
          <a
            class="c17"
            href="/todo/B"
          >
            Todo B
          </a>
          <div
            class="c18"
          >
            1. 1. 1970
          </div>
          <div
            class="c19"
            title="Hotovo"
          >
            <i
              class="c9"
            >
              ✔
            </i>
          </div>
          <div
            class=""
          >
            <a
              class="c8"
              href="/todo/B"
            >
              <i
                class="c9"
              >
                ✎
              </i>
              <span
                class="c10"
              >
                Upravit
              </span>
            </a>
          </div>
          <div
            class=""
          >
            <button
              class="c12 c20"
              type="button"
            >
              <i
                class="c9"
              >
                ✖
              </i>
              <span
                class="c10"
              >
                Smazat
              </span>
            </button>
          </div>
        </div>
      </li>
      <li
        class="c15"
      >
        <div
          class="c16"
        >
          <div
            class=""
          >
            <div
              class="c3"
            >
              <label
                class="c4"
                for="checkbox-7"
              >
                <input
                  class="c5"
                  id="checkbox-7"
                  type="checkbox"
                />
              </label>
            </div>
          </div>
          <a
            class="c17"
            href="/todo/C"
          >
            Todo B
          </a>
          <div
            class="c18"
          >
            1. 1. 1970
          </div>
          <div
            class="c19"
            title=""
          >
            <i
              class="c9"
            >
              -
            </i>
          </div>
          <div
            class=""
          >
            <a
              class="c8"
              href="/todo/C"
            >
              <i
                class="c9"
              >
                ✎
              </i>
              <span
                class="c10"
              >
                Upravit
              </span>
            </a>
          </div>
          <div
            class=""
          >
            <button
              class="c12 c20"
              type="button"
            >
              <i
                class="c9"
              >
                ✖
              </i>
              <span
                class="c10"
              >
                Smazat
              </span>
            </button>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`components/todo-list/TodoList should render empty list 1`] = `
.c9 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c12 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c13 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c13:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c13:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c13:disabled {
  color: #c0c0c0;
}

.c8 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-left: 5px;
  background: #f0f0f0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c8:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c8:disabled {
  color: #c0c0c0;
}

.c10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 18px;
  line-height: 30px;
  font-weight: normal;
  margin: 0;
}

.c14 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
}

.c6 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #c0c0c0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: 30px;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  width: 20px;
  height: 20px;
  border: 2px solid #c0c0c0;
  color: #c0c0c0;
  background: #fff;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c5 {
  position: absolute;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
  opacity: 0;
  outline: none;
}

.c2 {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  z-index: 1000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: calc(30px + 1px);
  border-bottom: 1px solid #c0c0c0;
  background: #f0f0f0;
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c11 {
  position: relative;
  display: inline-block;
}

.c1 {
  margin-bottom: 1px;
}

.c15 {
  padding-top: 10px;
}

<div>
  <div
    class=""
  >
    <h1
      class="c0 c1"
    >
      Úkoly
    </h1>
    <div
      class="c2"
    >
      <div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="checkbox-9"
          >
            <input
              class="c5"
              disabled=""
              id="checkbox-9"
              type="checkbox"
            />
          </label>
          <label
            class="c6"
            for="checkbox-9"
          >
            Vybrat vše
          </label>
        </div>
      </div>
      <div
        class="c7"
      />
      <div>
        <a
          class="c8"
          href="/todo/create"
        >
          <i
            class="c9"
          >
            +
          </i>
          <span
            class="c10"
          >
            Vytvořit
          </span>
        </a>
      </div>
      <div>
        <div
          class="c11"
        >
          <button
            class="c12 c13"
            disabled=""
            type="button"
          >
            <span
              class="c10"
            >
              Nejnovější
            </span>
            <i
              class="c9"
              style="color: rgb(192, 192, 192);"
            >
              ˅
            </i>
          </button>
        </div>
      </div>
    </div>
    <p
      class="c14 c15"
    >
      Nebyly nalezeny žádné úkoly.
    </p>
  </div>
</div>
`;
